JotunnLib/Managers/DungeonManager.cs

```diff
@@ -0,0 +1,372 @@
+            if (dg == null)
+            {
+                Logger.LogError($"Cannot find DungeonGenerator component in prefab {prefab.name}.");
+                throw new ArgumentException("Prefab must contain a DungeonGenerator component", "prefab");
+            }
+
+            DungeonGeneratorTheme dgt;
+            if (dg.gameObject.TryGetComponent(out dgt))
+            {
+                if (!string.IsNullOrEmpty(dgt.m_themeName) && dgt.m_themeName != themeName)
+                {
+                    Logger.LogWarning($"Overwriting existing theme name {dgt.m_themeName} with {themeName}.");
+                }
+            }
+            else
+            {
+                dgt = dg.gameObject.AddComponent<DungeonGeneratorTheme>();
+            }
+            dgt.m_themeName = themeName;
+            themeList.Add(themeName);
+
+            return true;
+        }
+
+
+        /// <summary>
+        ///     Adds a new environment prefab to be registered when <see cref="ZoneSystem.SetupLocations"/> runs.<br />
+        ///     If you intend to use a custom interior environment <see cref="Location.m_interiorEnvironment"/>, this method enables you 
+        ///     to provide a prefab with an appropriately configured <see cref="LocationList"/> containing atleast one
+        ///     <see cref="EnvSetup"/> within <see cref="LocationList.m_environments"/>.
+        /// </summary>
+        /// <param name="assetBundle">The <see cref="AssetBundle"/> containing the prefab.</param>
+        /// <param name="prefabName">The name of the prefab to register.</param>
+        public void RegisterEnvironment(AssetBundle assetBundle, string prefabName)
+        {
+            if (assetBundle == null)
+            {
+                throw new ArgumentException("Cannot be null", nameof(assetBundle));
+            }
+            if (string.IsNullOrEmpty(prefabName))
+            {
+                throw new ArgumentException("Cannot be null or empty", nameof(prefabName));
+            }
+            var env = assetBundle.LoadAsset<GameObject>(prefabName);
+            loadedEnvironments.Add(prefabName, env);
+        }
+
+        private void GenerateHashes()
+        {
+            hashToName.Clear();
+
+            foreach (CustomRoom room in Rooms.Values)
+            {
+                int stableHashCode = room.Prefab.name.GetStableHashCode();
+                if (hashToName.ContainsKey(stableHashCode))
+                {
+                    Logger.LogWarning($"Room {room.Name} is already registered with hash {stableHashCode}");
+                }
+                else
+                {
+                    hashToName.Add(stableHashCode, room.Name);
+                }
+            }
+        }
+
+        private void OnZoneSystemSetupLocations()
+        {
+            // TODO - unsure if cache clearing is still necessary for resolving mocks of these assets
+            //ClearPrefabCache(typeof(Material));
```

Comment:
it's fine to preemptively clear the Material cache. It's not an expensive operation unlike clearing the GameObject cache, because there are a lot more GameObjects than materials